File: 1509-minimum-difference-between-largest-and-smallest-value-in-three-moves/1509-minimum-difference-between-largest-and-smallest-value-in-three-moves.py
class Solution:
    def minDifference(self, nums: List[int]) -> int:
        n = len(nums)
        if n <= 4:
            return 0
        
        pq_max = []
        pq_min = []
        
        for num in nums:
            heapq.heappush(pq_min, num)
            heapq.heappush(pq_max, -num)
            if len(pq_max) > 4:
                heapq.heappop(pq_max)
            if len(pq_min) > 4:
                heapq.heappop(pq_min)
        
        l, r = 3, n - 4
        
        while pq_max:
            nums[l] = -heapq.heappop(pq_max)
            nums[r] = heapq.heappop(pq_min)
            l -= 1
            r += 1
        
        return min(nums[n-1] - nums[3], nums[n-2] - nums[2], nums[n-3] - nums[1], nums[n-4] - nums[0])






# class Solution:
#     def minDifference(self, nums: List[int]) -> int:
#         n = len(nums)
#         if n <= 4:
#             return 0
        
#         nums.sort()
        
#         min_diff = float('inf')
#         min_diff = min(min_diff, nums[n-1] - nums[3])
#         min_diff = min(min_diff, nums[n-2] - nums[2])
#         min_diff = min(min_diff, nums[n-3] - nums[1])
#         min_diff = min(min_diff, nums[n-4] - nums[0])
        
#         return min_diff

File: LeetCode/1509-minimum-difference-between-largest-and-smallest-value-in-three-moves/1509-minimum-difference-between-largest-and-smallest-value-in-three-moves.cpp
class Solution {
public:
    int minDifference(vector<int>& nums) {
        int n = nums.size();
        if (n <= 4) {
            return 0;
        }
        priority_queue<int> pq_max;
        priority_queue<int, vector<int>, greater<int>> pq_min;
        for (int num : nums) {
            pq_min.push(num);
            pq_max.push(num);
            if (pq_max.size() > 4) {
                pq_max.pop();
            }
            if (pq_min.size() > 4) {
                pq_min.pop();
            }
        }
        int l = 3, r = n - 4;
        while (!pq_max.empty()) {
            nums[l--] = pq_max.top();
            pq_max.pop();
            nums[r++] = pq_min.top();
            pq_min.pop();
        }
        return min({nums[n-1] - nums[3], nums[n-2] - nums[2], nums[n-3] - nums[1], nums[n-4] - nums[0]});
    }
};



// class Solution {
// public:
//     int minDifference(vector<int>& nums) {
//         int n = nums.size();
//         if (n <= 4) {
//             return 0;
//         }
//         sort(nums.begin(), nums.end());
//         return min({nums[n-1] - nums[3], nums[n-2] - nums[2], nums[n-3] - nums[1], nums[n-4] - nums[0]});
//     }
// };